### PR TITLE
Stdin support

### PR DESCRIPTION
**Pull request by johannhof**
I was trying to get rustfmt work with https://github.com/Chiel92/vim-autoformat and found to my surprise that there seems to be no support for receiving a string on stdin and outputting a simple formatting to stdout.

This is actually the default behavior of most formatting tools, like `gofmt`, and really vital for IDE and tooling integration. I'm sorry if I'm misunderstanding something and rustfmt already supports it somehow. Feel free to close in that case!

Most of the existing code seems to be built with a multiple-file codemap in mind, making it hard to whip up a perfect solution in trivial time. My approach is trying to get the job done without breaking existing behavior.

tl;dr
- `rustfmt` will now read from stdin if no file was specified and format that code without walking any other files
- a new write-mode `plain` was introduced to print to stdout without any fluff

### CODE REVIEW

File: src/bin/rustfmt.rs [hunk 1]
```diff
@@ -138,6 +152,32 @@ fn determine_operation<I>(opts: &Options, args: I) -> Operation
         return Operation::Help;
     }
 
+    // if no file argument is supplied, read from stdin
+    if matches.free.len() != 1 {
```
Comment (nrc) [inline src/bin/rustfmt.rs]: I'd prefer `== 0` here - I can imagine we might want to make > 1 args work some day, or give an error message rather than reading from stdin

File: src/bin/rustfmt.rs [hunk 2]
```diff
@@ -138,6 +152,32 @@ fn determine_operation<I>(opts: &Options, args: I) -> Operation
...
+            Some(mode) => {
+                match mode.parse() {
+                    Ok(WriteMode::Plain) => (),
+                    _ => return Operation::InvalidInput("Using stdin requires write-mode to be \
+                                                         plain"
+                                                            .into()),
+                }
+            }
+            _ => (),
+        }
```
Comment (nrc) [inline src/bin/rustfmt.rs]: Can we just ignore the write mode and always use plain? Would allow a bit more flexibility

File: src/filemap.rs [hunk 3]
```diff
@@ -100,6 +100,11 @@ fn write_file(text: &StringBuffer,
             let file = try!(File::create(&filename));
             try!(write_system_newlines(file, text, config));
         }
+        WriteMode::Plain => {
+            let stdout = stdout();
+            let stdout_lock = stdout.lock();
```
Comment (nrc) [inline src/filemap.rs]: style nit: recycle the name `stdout` here.

File: src/lib.rs [hunk 4]
```diff
@@ -386,6 +389,33 @@ pub fn fmt_lines(file_map: &mut FileMap, config: &Config) -> FormatReport {
...
+    // do the actual formatting
+    let mut visitor = FmtVisitor::from_codemap(&parse_session, config, Some(mode));
+    visitor.format_separate_mod(&krate.module, path);
+
+    // append final newline
+    visitor.buffer.push_str("\n");
+    file_map.insert(path.to_owned(), visitor.buffer);
+
+    return file_map;
+}
```
Comment (nrc) [inline src/lib.rs]: It would be nice (but not essential) to try to share more code between this function and `format`.

**Comment by nrc** [comment]
This is great, thanks for the PR! A few minor comments inline.

File: src/bin/rustfmt.rs (hunk 2 above)
Comment (johannhof) [inline src/bin/rustfmt.rs]: I was looking at how `gofmt` does it
```
echo "1+1" | gofmt -w                                                                                                                                                         
> error: cannot use -w with standard input
```
hence I decided to give an error message. It has upsides and downsides. It's more graceful but might surprise some users.

But I actually agree with you, the only thing I would do then is add a warning in the `--help` usage notes that stdin will always print using plain.

File: src/lib.rs (hunk 4 above)
Comment (johannhof) [inline src/lib.rs]: yeah I agree, let me see what I can come up with

Comment (johannhof) [inline src/lib.rs]: So after looking at it again, the only code that is actually shared is 
```
    // Suppress error output after parsing.
    let emitter = Box::new(EmitterWriter::new(Box::new(Vec::new()), None));
    parse_session.span_diagnostic.handler = Handler::with_emitter(false, emitter);
```
which isn't worth making a new function IMO. I would much rather try to tackle eliminating the `FileMap` in another PR (as mentioned in the FIXME). Then it will become more apparent that the code in `format_string` is really different from `format`.

So I'd leave it like this for now, if that's ok with you.

**Comment by johannhof** [comment]
@nrc do you like your commits squashed or stacked? :cocktail:

**Comment by nrc** [comment]
Thanks for the changes!

If there's a lot of commits then a squash is good, but two is fine.
